subsys/shell/shell.c

```diff
@@ -1259,6 +1256,8 @@ int shell_start(const struct shell *shell)
 		return -ENOTSUP;
 	}
 
+	k_mutex_lock(&shell->ctx->wr_mtx, K_FOREVER);
```

Comment:
can you move this lock to `shell_internal_fprintf` function? I noticed that internal fprintf is actually not protected so if `shell_fprintf` is called from other context than shell command there is a chance that it will clash.

Reply:
That is likely a better place for the lock.  I will have a look at that tomorrow.

Reply:
@nordic-krch I had a closer look and it appears that shell_fprintf() and shell_internal_fprintf() both call shell_internal_vfprintf().  So, perhaps the best place for the mutex is shell_internal_vfprintf()?

Reply:
All of the printf's appear to converge to shell_fprintf_fmt() in shell_fprintf.c.  We do not have the same mutex because we do not have the shell struct passed down to here.  So, we would need to create another one.  Do you see any problem with that?

Reply:
No, cannot put it there ... that is a const struct.

Reply:
ok, i think that your initial proposal is actually ok.

Whole shell processing is done under mutex lock:
https://github.com/zephyrproject-rtos/zephyr/blob/ef4d62749c9688e60206a9cbc299dcacf0f59767/subsys/shell/shell.c#L1189

It is unlocked only when user command handler is called to allow printing from that context. If `shell_print` is called from another context then this mutex will protect access. What is not protected is the start portion and your initial patch added it.

Reply:
@nordic-krch 
Function `shell_internal_fprintf` is not protected by the lock on purpose. Please do not lock it :)

Reply:
@jakub-uC agree, my initial proposal was wrong. However, shell start shall be locked.

Reply:
@nordic-krch 

> It is unlocked only when user command handler is called to allow printing from that context. If shell_print is called from another context then this mutex will protect access.

I am attempting to print a banner to the shell when my board boots.  Here is a snippet of code in main.c where I wait for the DTR signal from the USB_CDC and then print the banner.  There are a couple calls to shell_fprintf in there:
```
	{
		u8_t i = 0;
		u32_t dtr = 0U;
		
		LOG_INF("Waiting for shell ...");
		k_timer_start(&loop_timeout_timer, K_MSEC(WAIT_FOR_KEYPRESS_PERIOD), K_MSEC(0));
		while (true) {
			uart_line_ctrl_get(dev, UART_LINE_CTRL_DTR, &dtr);
			if (dtr) {
				ret = uart_line_ctrl_set(dev, UART_LINE_CTRL_DCD, 1);
				ret = uart_line_ctrl_set(dev, UART_LINE_CTRL_DSR, 1);
				uart_irq_tx_enable(sh_uart->ctrl_blk->dev);
				LOG_DBG("shell ready");
				shell_fprintf(sh, SHELL_VT100_COLOR_BLUE, "%s", STRING_SCAN_LOGO);
				shell_fprintf(sh, SHELL_VT100_COLOR_BLUE, "%s", STRING_WITAP_LOGO);
				break;
			} else {
				if ( k_timer_status_get(&loop_timeout_timer) > 0 ) {
					// if we time out, detach the USB
					LOG_DBG("shell wait timed out");
					k_delayed_work_submit(&usb_disconnect_work, K_MSEC(1000));
					break;
				}
				if ((i++ % 30) == 0) {
					wdt_feed(wdt, wdt_channel_id);
					led_add(&animation_wait);
				}
				k_sleep(K_MSEC(100));
			}
		}
	}
```
Did you intend your comment to mean calling shell_fprintf() from anything but the shell context as shown in the example above is forbidden?

Reply:
@jakub-uC Is it possible that not protecting shell_fprintf_fmt() can cause a problem because of shell_pend_on_txdone()?  When I was debugging this issue I encountered problems when shell_pend_on_txdone() waits for an event that never happens.  One example of this is encountered when starting the shell.  I have noticed other situations where shell_pend_on_txdone() pends forever.

Reply:
I compiled an experiment where I put a lock here:
```
void shell_fprintf_fmt(const struct shell_fprintf *sh_fprintf,
		       const char *fmt, va_list args)
{
	k_mutex_lock(&sh_fprintf->ctrl_blk->print_mtx, K_FOREVER);
#if !defined(CONFIG_NEWLIB_LIBC) && !defined(CONFIG_ARCH_POSIX)
	(void)z_prf(out_func, (void *)sh_fprintf, (char *)fmt, args);
#else
	z_vprintk(out_func, (void *)sh_fprintf, fmt, args);
#endif

	if (sh_fprintf->ctrl_blk->autoflush) {
		shell_fprintf_buffer_flush(sh_fprintf);
	}
	k_mutex_unlock(&sh_fprintf->ctrl_blk->print_mtx);
}
```
I declare the lock in the shell_fprintf_control_block struct and it is initialized in instance_init().

Reply:
@sslupsky : I've started working on that, sorry for the delay.
Basically, the shell allows using only functions /macros that are protected by the mutex like `shell_fprintf`. It is calling `shell_vfprintf` that has the mutex.

Function `shell_fprintf_fmt` can be called internally by the shell because it assumes that the mutex taken by the shell thread.  This function does not belong to shell API and shall not be used. You cannot add mutex there.